pgrx/src/spi/query.rs

```diff
@@ -41,32 +41,84 @@ pub trait Query<'conn>: Sized {
     ) -> SpiResult<SpiCursor<'conn>>;
 }
 
-impl<'conn> Query<'conn> for &String {
-    type Arguments = Option<Vec<(PgOid, Option<pg_sys::Datum>)>>;
-
-    fn execute(
+/// A trait representing a query which can be prepared.
+pub trait PreparableQuery<'conn>: Query<'conn> {
```

Comment:
This probably seems obvious but can you explain how this distinguishes from Query?

Reply:
To distinguish from the prepared statements which should not be used by the prepare methods of the SPI client.

Reply:
Hm, not entirely sure what the benefit is?

Reply:
This is probably obvious to you, it's just been ages since I've looked at the Query API and my main dread is it being cemented further, because I would rather it... not exist, tbh. It's... overly cutesy.

Reply:
The problem that trait solves is prohibiting `PreparedStatement<'conn>` from being used with `prepare` and `prepare_mut` methods of the `SqlClient<'conn>`, so the following code will result in a compiler error:

```rust
let conn = SpiConnection::connect()?;
let client = conn.client();

let once = client.prepare(c"SELECT 1", None)?;
let twice = client.prepare(once, None);
//                 ------- ^^^^ the trait `PreparableQuery<'_>` is not implemented for `PreparedStatement<'_>`
//                              |
//                              required by a bound introduced by this call
```

Reply:
Ah.

Can you add a UI test to verify it has a compiler error then?

Reply:
I played with the CI and observed that my UI test fails only for PostgreSQL versions 12, 13 and 14, but not for others. That's totally weird. See #1870.

Reply:
...huh.

Reply:
...oh, I remember. They introduced a type named String in PostgreSQL. Can you defer the test to opening an issue instead?

Reply:
This one, right? https://doxygen.postgresql.org/structString.html

Reply:
Yes.